csv/parse.ts

```diff
@@ -333,26 +347,139 @@ export function parse(input: string): string[][];
 /**
  * Parses CSV string into an array of objects or an array of arrays of strings.
  *
- * If `column` or `skipFirstRow` option is provided, it returns an array of
+ * If `columns` or `skipFirstRow` option is provided, it returns an array of
  * objects, otherwise it returns an array of arrays of string.
  *
- * @example Usage
+ * @example Don't skip first row with `skipFirstRow: false`
  * ```ts
  * import { parse } from "@std/csv/parse";
- * import { assertEquals } from "@std/assert";
+ * import { assertEquals } from "@std/assert/equals";
+ * import { assertType, IsExact } from "@std/testing/types"
  *
  * const string = "a,b,c\nd,e,f";
+ * const result = parse(string, { skipFirstRow: false });
+ *
+ * assertEquals(result, [["a", "b", "c"], ["d", "e", "f"]]);
+ * assertType<IsExact<typeof result, string[][]>>(true);
+ * ```
+ *
+ * @example Skip first row with `skipFirstRow: true`
+ * ```ts
+ * import { parse } from "@std/csv/parse";
+ * import { assertEquals } from "@std/assert/equals";
+ * import { assertType, IsExact } from "@std/testing/types"
+ *
+ * const string = "a,b,c\nd,e,f";
+ * const result = parse(string, { skipFirstRow: true });
+ *
+ * assertEquals(result, [{ a: "d", b: "e", c: "f" }]);
+ * assertType<IsExact<typeof result, Record<string, string | undefined>[]>>(true);
+ * ```
+ *
+ * @example Specify columns with `columns` option
+ * ```ts
+ * import { parse } from "@std/csv/parse";
+ * import { assertEquals } from "@std/assert/equals";
+ * import { assertType, IsExact } from "@std/testing/types"
+ *
+ * const string = "a,b,c\nd,e,f";
+ * const result = parse(string, { columns: ["x", "y", "z"] });
+ *
+ * assertEquals(result, [{ x: "a", y: "b", z: "c" }, { x: "d", y: "e", z: "f" }]);
+ * assertType<IsExact<typeof result, Record<"x" | "y" | "z", string>[]>>(true);
+ * ```
+ *
+ * @example Specify columns with `columns` option and skip first row with
+ * `skipFirstRow: true`
+ * ```ts
+ * import { parse } from "@std/csv/parse";
+ * import { assertEquals } from "@std/assert/equals";
+ * import { assertType, IsExact } from "@std/testing/types"
+ *
+ * const string = "a,b,c\nd,e,f";
+ * const result = parse(string, { columns: ["x", "y", "z"], skipFirstRow: true });
+ *
+ * assertEquals(result, [{ x: "d", y: "e", z: "f" }]);
+ * assertType<IsExact<typeof result, Record<"x" | "y" | "z", string>[]>>(true);
+ * ```
+ *
+ * @example TSV (tab-separated values) with `separator: "\t"`
+ * ```ts
+ * import { parse } from "@std/csv/parse";
+ * import { assertEquals } from "@std/assert/equals";
+ *
+ * const string = "a\tb\tc\nd\te\tf";
+ * const result = parse(string, { separator: "\t" });
+ *
+ * assertEquals(result, [["a", "b", "c"], ["d", "e", "f"]]);
+ * ```
+ *
+ * @example Trim leading space with `trimLeadingSpace: true`
```

Comment:
Is there a reason there's an option to trim only leading spaces? Why not just trim the end too using `String.prototype.trim()`?

Reply:
I'm not sure either, but according to the head comment of parse.ts, many parts of the implementation was ported from Go which has the same option.
https://github.com/golang/go/blob/go1.12.5/src/encoding/csv/reader.go#L136

Reply:
I can't think of a reason to only trim the start... I suggest changing to `trim: boolean`. Not a strong opinion. What do we think?

Reply:
Asked ChatGPT, it answered some application may add leading whitespaces when exporting to CSV. I don't know if this is true at all 😄 Anyway `trim: boolean` should be useful in some scenarios and I think we could add that after 1.0 without breaking things?

Reply:
I think trailing spaces are usually considered as a part of value of the last column (For example, excel handles it in that way)

Reply:
I couldn't find the case when this returns `Record<string, string | undefined>`, but I  also noticed that `fieldsPerRecord` option doesn't seem working as described.

It says:
> If negative, no check is made and records may have a variable number of fields.

But `parse` doesn't seem allowing wrong number of fields even when I passed negative number to this option. Maybe this current state is related to it?

Reply:
> But parse doesn't seem allowing wrong number of fields even when I passed negative number to this option. Maybe this current state is related to it?

Oh that's a good find. I'll add a test case and fix it